### DISCUSSION TITLE
.MAN data different from reality.

### DISCUSSION THREAD

**Discussion by Skino1337** [GothicKit/ZenKit#92] [Q&A]
Hello there, when reading the ".MAN" file I get incorrect position and rotation data, here is the position data from the first (0 index) bone:
![script](https://github.com/GothicKit/ZenKit/assets/5137675/f70105e1-cdfd-4f0e-8100-775d2bcb62d7)

But the position data from the first bone (0 index) is already in blender, you can see from the curves that the data is different:
![blender](https://github.com/GothicKit/ZenKit/assets/5137675/2ddb0c24-b294-433e-99b9-626019c2d8b7)

The data in blender is correct, what's wrong with the data from the ".MAN" file?

code snipped:
```
            for bone_index, bone_index_key in enumerate(model_animation.node_indices):
                bone_name = model_hierarchy_data[model_animation.checksum]['bones'][bone_index_key]
                for frame_index in range(model_animation.frame_count):
                    pos = model_animation.samples[bone_index + frame_index].position
                    pos_parsed = parse_quat_or_vec(pos)

                    rot = model_animation.samples[bone_index + frame_index].rotation
                    rot_parsed = parse_quat_or_vec(rot)

                    model_animation_data[node.name][bone_name]['pos'].append(pos_parsed)
                    model_animation_data[node.name][bone_name]['rot'].append(rot_parsed)
```

and files:
[rep.zip](https://github.com/user-attachments/files/16054733/rep.zip)

**Comment by lmichaelis** [comment]
Hm I don't know what the problem is here. Generally the animation data is correct (see OpenGothic, for example) but it is possible that it's a bit quirly, especially with the start position. Now in this case, I can imagine that the values in Blender's animation editor are different from the values stored in the MAN. It is important to notice the the values in the MAN are independent, meaning the position of every bone is absolute, not relative to its parent bone.

Seeing the code for `parse_quat_or_vec` would probably help :)

**Comment by Skino1337** [comment]
```
def parse_quat_or_vec(data):
    # print(data)
    data = str(data).replace('Quat', '').replace('Vec3f', '')
    data = data.replace('x=', '').replace('y=', '').replace('z=', '').replace('w=', '')
    data = data.replace('(', '').replace(')', '').replace(',', '')
    data = data.split()
    data = [float(f) for f in data]
    data = [round(f, 2) for f in data]
    # print(data)

    return data
```

_It is important to notice the the values in the MAN are independent, meaning the position of every bone is absolute, not relative to its parent bone._ Good info, but its first bone.

**Reply by lmichaelis** [reply]
Well that's a weird way to access that data. You can just `data.x`, `data.y`, `data.z` and if `isinstance(data, Quat)` then you can also access `data.w`.

**Reply by Skino1337** [reply]
![image](https://github.com/GothicKit/ZenKit/assets/5137675/52c8bcd2-154a-4ca2-a7ef-bf7b646f65ac)

**Reply by lmichaelis** [reply]
Try it :) it doesn't show because it's a `ctypes` struct which has dynamic fields:
![image](https://github.com/GothicKit/ZenKit/assets/20187490/9b1eaa44-26d5-4ca2-acbc-4de2fcb08903)

**Reply by Skino1337** [reply]
thx, its work, but not help for my global question. =(

**Comment by Skino1337** [comment]
found smth, i use wrong iteration direction...
